src/Common/Form/FileType.php

```diff
@@ -173,33 +189,101 @@ function ($optionName) use ($options, &$view) {
+    private function getMaxFileSizeServerValue(): ?int
     {
         $uploadMaxFileSize = ini_get('upload_max_filesize');
+
         if ($uploadMaxFileSize === false) {
             return null;
         }
 
         // reformat if defined as an integer
         if (is_numeric($uploadMaxFileSize)) {
-            return $uploadMaxFileSize / 1024 . 'MB';
+            return $uploadMaxFileSize;
         }
 
         // reformat if specified in kB
         if (mb_strtoupper(mb_substr($uploadMaxFileSize, -1)) === 'K') {
-            return mb_substr($uploadMaxFileSize, 0, -1) . 'kB';
+            return mb_substr($uploadMaxFileSize, 0, -1) * 1000;
         }
 
         // reformat if specified in MB
         if (mb_strtoupper(mb_substr($uploadMaxFileSize, -1)) === 'M') {
-            return $uploadMaxFileSize . 'B';
+            return mb_substr($uploadMaxFileSize, 0, -1) * 1000 * 1000;
         }
 
         // reformat if specified in GB
         if (mb_strtoupper(mb_substr($uploadMaxFileSize, -1)) === 'G') {
-            return $uploadMaxFileSize . 'B';
+            return mb_substr($uploadMaxFileSize, 0, -1) * 1000 * 1000 * 1000;
+        }
+
+        return null;
+    }
+
+    private function getMaxFileSizeConstraintValue(string $fileClass): ?int
+    {
+        // use the metaData to find the file data
+        /** @var ClassMetadata $metaData */
+        $metaData = $this->validator->getMetadataFor($fileClass);
+        $members = $metaData->getPropertyMetadata('file');
+
+        // the constraints can be found in the property meta data members
+        foreach ($members as $member) {
+            $constraints = $member->getConstraints();
+
+            if (count($constraints) === 1) {
+                /** @var File $fileConstraint */
+                $fileConstraint = $constraints[0];
+
+                if ($fileConstraint instanceof File) {
+                    return $fileConstraint->maxSize;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private function getUploadMaxFileSize(string $fileClass): ?string
+    {
+        $constraintValue = $this->getMaxFileSizeConstraintValue($fileClass);
+        $serverValue = $this->getMaxFileSizeServerValue();
+
+        if (!is_numeric($constraintValue) && !is_numeric($serverValue)) {
+            return null;
+        }
+
+        // return the server value if the constraint value is to high
+        if (is_numeric($constraintValue) && is_numeric($serverValue)) {
+            if ($constraintValue < $serverValue) {
+                return $this->prettyPrintFileSize($constraintValue);
+            }
+
+            return $this->prettyPrintFileSize($serverValue);
+        }
+
+        if (is_numeric($constraintValue)) {
+            return $this->prettyPrintFileSize($constraintValue);
+        }
+
+        return $this->prettyPrintFileSize($serverValue);
+    }
+
+    private function prettyPrintFileSize(int $fileSize): string
```

Comment:
Can't this function be placed in somewhere else since it's also be being used in ImageType?